File: server/battlesnake.go
package main

import (
	"database/sql"

	"github.com/gin-gonic/gin"
)

type Player struct {
	Username string `json:"username"`
	ID string `json:"id"`
}

type NewRoom struct {
	Name string `json:"name" binding:"required,min=1,max=32"`
	Password string `json:"password" binding:"max=32"`
}

type Room struct {
	Name string `json:"name"`
	ID string `json:"id"`
	Players []Player `json:"players"`
	Password string `json:"-"`
	HasPassword bool `json:"password"`
}

var rooms []Room

func getRooms(c *gin.Context) {
	c.JSON(200, rooms)
}

func createRoom(c *gin.Context) {
	var n NewRoom
	if err := c.BindJSON(&n); err != nil {
		c.AbortWithError(400, err)
	}

	var r Room
	r.Name = n.Name
	r.Password = n.Password

	r.ID = id()
	r.HasPassword = len(r.Password) > 0
	rooms = append(rooms, r)

	c.String(201, r.ID)
}

func battlesnake(g *gin.RouterGroup, _ *sql.DB) {
	battlesnake := g.Group("/battlesnake")

	battlesnake.GET("/rooms", getRooms)
	battlesnake.POST("/rooms", createRoom)
}

File: server/routes/battlesnake.mjs
import { Router } from 'express';

import createRoom from '../battlesnake/room.mjs';
import createPlayer from '../battlesnake/player.mjs';

export const router = Router();


let rooms = new Map();

// routes
router.get('/rooms', (req, res) => {
	const sanitizedRooms = [];
	for (const room of rooms.values()) {
		sanitizedRooms.push(room.sanitized());
	}
	res.json(sanitizedRooms);
});

router.post('/rooms', (req, res) => {
	if (!req.body.name) {
		res.sendStatus(400);
		return;
	}

	req.body.name = req.body.name.slice(0, 32);
	req.body.password = req.body.password?.slice(0, 32);

	const newRoom = createRoom(req.body.name, req.body.password);
	newRoom.onclose = () => {
		rooms.delete(newRoom.id);
	};
	rooms.set(newRoom.id, newRoom)

	res.status(201).type('txt').send(newRoom.id);
});

// io
export function ioInit(io) {
	io.on('connection', socket => {
		socket.on('join', (roomID, username, cb) => {
			const room = rooms.get(roomID);
			if (!room) {
				cb({
					status: 'err',
					err: 'notfound',
				});
				return;
			}

			const player = createPlayer(username, socket.id, socket);
			if (!room.io) {
				// player is host, initialize room io
				room.ioInit(io.in(room.id));
				player.isHost = true;
			}
			if (!room.add(player)) {
				cb({
					status: 'err',
					err: 'full',
				});
				return;
			}

			cb({
				status: 'ok',
				room: room.sanitized(),
				isHost: player.isHost,
			});
		});
	});
}
